jerry-core/ecma/operations/ecma-comparison.c

```diff
@@ -391,14 +391,11 @@ ecma_op_abstract_relational_compare (ecma_value_t x, /**< first operand */
       ecma_number_t nx;
       ecma_number_t ny;
 
-      if (ECMA_IS_VALUE_ERROR (ecma_op_to_number (px, &nx)))
+      if (ECMA_IS_VALUE_ERROR (ecma_op_to_number (px, &nx))
+          || ECMA_IS_VALUE_ERROR (ecma_op_to_number (py, &ny)))
       {
-        return ECMA_VALUE_ERROR;
-      }
-
-      if (ECMA_IS_VALUE_ERROR (ecma_op_to_number (py, &ny)))
-      {
-        return ECMA_VALUE_ERROR;
+        ret_value = ECMA_VALUE_ERROR;
```

Comment:
This assignment would be no longer needed if `ret_value` would be initialized to `ECMA_VALUE_ERROR` at the variable declaration.

Reply:
I'm not 100% sure if it will be correct. There is a ret_value == ECMA_VALUE_EMPTY later in the code.
I can't see at first sight if it expects the init empty value or an empty value from a later assignment.

Reply:
I checked your idea, it is definitely wrong and would cause assertion on tests/jerry/es.next/bigint4.js test:
ICE: Assertion 'jcontext_has_pending_exception ()' failed at /home/oszi/jerryscript/jerry-core/vm/vm.c(vm_loop):4556.
Error: ERR_FAILED_INTERNAL_ASSERTION

Reply:
This idea is definitely **NOT** wrong only some bigint comparison adjustment was required.

```diff
diff --git a/jerry-core/ecma/operations/ecma-comparison.c b/jerry-core/ecma/operations/ecma-comparison.c
index 988bca810..8bd0f198c 100644
--- a/jerry-core/ecma/operations/ecma-comparison.c
+++ b/jerry-core/ecma/operations/ecma-comparison.c
@@ -355,7 +355,6 @@ ecma_op_abstract_relational_compare (ecma_value_t x, /**< first operand */
                                      ecma_value_t y, /**< second operand */
                                      bool left_first) /**< 'LeftFirst' flag */
 {
-  ecma_value_t ret_value = ECMA_VALUE_EMPTY;
 
   /* 1., 2. */
   ecma_value_t prim_first_converted_value = ecma_op_to_primitive (x, ECMA_PREFERRED_TYPE_NUMBER);
@@ -376,6 +375,7 @@ ecma_op_abstract_relational_compare (ecma_value_t x, /**< first operand */
 
   const bool is_px_string = ecma_is_value_string (px);
   const bool is_py_string = ecma_is_value_string (py);
+  ecma_value_t ret_value = ECMA_VALUE_ERROR;
 
   if (!(is_px_string && is_py_string))
   {
@@ -394,7 +394,6 @@ ecma_op_abstract_relational_compare (ecma_value_t x, /**< first operand */
       if (ECMA_IS_VALUE_ERROR (ecma_op_to_number (px, &nx))
           || ECMA_IS_VALUE_ERROR (ecma_op_to_number (py, &ny)))
       {
-        ret_value = ECMA_VALUE_ERROR;
         goto end;
       }
 
@@ -498,12 +497,12 @@ ecma_op_abstract_relational_compare (ecma_value_t x, /**< first operand */
           if (ret_value == ECMA_VALUE_FALSE)
           {
             ret_value = ECMA_VALUE_UNDEFINED;
+            goto end;
           }
           else
           {
             compare_result = ecma_bigint_compare_to_bigint (px, ret_value);
             ecma_free_value (ret_value);
-            ret_value = ECMA_VALUE_EMPTY;
           }
         }
       }
@@ -512,13 +511,13 @@ ecma_op_abstract_relational_compare (ecma_value_t x, /**< first operand */
         ecma_number_t ny;
         if (ECMA_IS_VALUE_ERROR (ecma_op_to_number (py, &ny)))
         {
-          ret_value = ECMA_VALUE_ERROR;
           goto end;
         }
 
         if (ecma_number_is_nan (ny))
         {
           ret_value = ECMA_VALUE_UNDEFINED;
+          goto end;
         }
         else
         {
@@ -526,15 +525,12 @@ ecma_op_abstract_relational_compare (ecma_value_t x, /**< first operand */
         }
       }
 
-      if (ret_value == ECMA_VALUE_EMPTY)
+      if (invert_result)
       {
-        if (invert_result)
-        {
-          compare_result = -compare_result;
-        }
-
-        ret_value = ecma_make_boolean_value (compare_result < 0);
+        compare_result = -compare_result;
       }
+
+      ret_value = ecma_make_boolean_value (compare_result < 0);
     }
 #endif /* ENABLED (JERRY_BUILTIN_BIGINT) */
   }

```

Reply:
Of course, it can be done with refactoring unrelated codes. But please keep in mind that the goal of this PR is ONLY to fix a memory leak, not to reach make ecma_op_abstract_relational_compare smallest code size. I support you to make it more optimal, but please do it in a follow-up PR.

Reply:
Conclusion: Since this change is not related to the PR, I'm waiting for the follow-up patch.